File: java/client/src/main/java/glide/api/MonitorClient.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api;

import com.google.protobuf.ByteString;
import connection_request.ConnectionRequestOuterClass.AuthenticationInfo;
import connection_request.ConnectionRequestOuterClass.ConnectionRequest;
import connection_request.ConnectionRequestOuterClass.NodeAddress;
import connection_request.ConnectionRequestOuterClass.ProtocolVersion;
import connection_request.ConnectionRequestOuterClass.TlsMode;
import glide.api.models.commands.MonitorMsg;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.ServerCredentials;
import glide.api.models.exceptions.ConnectionException;
import glide.internal.GlideNativeBridge;
import glide.internal.MonitorCallback;
import glide.managers.TlsConfigHelper;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import lombok.NonNull;

/**
 * A dedicated client for the MONITOR command. Opens its own connection to the Valkey server and
 * streams all commands processed by the server.
 *
 * <p>Standalone (non-cluster) connections only.
 *
 * <p>Supports two consumption modes:
 *
 * <ul>
 *   <li>Callback mode: supply a {@link Consumer} to {@link #create}; each message is delivered to
 *       the callback on a background thread.
 *   <li>Queue mode: use {@link #getMonitorMessage()} / {@link #tryGetMonitorMessage()} to poll.
 * </ul>
 *
 * @since Valkey 1.0.0.
 */
public class MonitorClient implements AutoCloseable {

    private final long nativeMonitorId;
    private final AtomicBoolean closed = new AtomicBoolean(false);
    private final BlockingQueue<MonitorMsg> queue;

    private MonitorClient(long nativeMonitorId, BlockingQueue<MonitorMsg> queue) {
        this.nativeMonitorId = nativeMonitorId;
        this.queue = queue;
    }

    /**
     * Creates a new {@link MonitorClient} for the given standalone configuration.
     *
     * @param config The standalone client configuration.
     * @return A new {@link MonitorClient} instance.
     * @throws ConnectionException if the native monitor client cannot be created.
     * @throws NullPointerException if {@code config} is null.
     * @throws IllegalArgumentException Not applicable - cluster configurations are rejected at
     *     compile time since this method only accepts {@link GlideClientConfiguration}.
     */
    public static MonitorClient create(@NonNull GlideClientConfiguration config)
            throws ConnectionException {
        return create(config, null);
    }

    /**
     * Creates a new {@link MonitorClient} with a callback for each monitor message.
     *
     * @param config The standalone client configuration.
     * @param callback Optional callback invoked for each message. If null, messages are queued.
     * @return A new {@link MonitorClient} instance.
     * @throws ConnectionException if the native monitor client cannot be created.
     */
    public static MonitorClient create(
            @NonNull GlideClientConfiguration config, Consumer<MonitorMsg> callback)
            throws ConnectionException {
        byte[] configBytes = serializeConfig(config);

        // Create the queue first so the JNI callback closure captures it directly.
        BlockingQueue<MonitorMsg> queue = new LinkedBlockingQueue<>();
        MonitorCallback jniCallback =
                (timestamp, db, clientAddr, command, argsJson) -> {
                    List<String> args = parseArgsJson(argsJson);
                    MonitorMsg msg = new MonitorMsg(timestamp, db, clientAddr, command, args);
                    if (callback != null) {
                        callback.accept(msg);
                    } else {
                        queue.offer(msg);
                    }
                };

        long monitorId = GlideNativeBridge.createMonitorClient(configBytes, jniCallback);
        if (monitorId == 0) {
            throw new ConnectionException("Failed to create native monitor client");
        }

        return new MonitorClient(monitorId, queue);
    }

    /**
     * Blocks until a monitor message is available.
     *
     * @return The next {@link MonitorMsg}, or {@code null} if the client is closed.
     * @throws InterruptedException if interrupted while waiting.
     */
    public MonitorMsg getMonitorMessage() throws InterruptedException {
        while (!closed.get()) {
            MonitorMsg msg = queue.poll(100, TimeUnit.MILLISECONDS);
            if (msg != null) return msg;
        }
        return null;
    }

    /**
     * Blocks until a monitor message is available or the timeout expires.
     *
     * @param timeoutMs Maximum wait time in milliseconds.
     * @return The next {@link MonitorMsg}, or {@code null} if timed out or closed.
     * @throws InterruptedException if interrupted while waiting.
     */
    public MonitorMsg getMonitorMessage(long timeoutMs) throws InterruptedException {
        if (closed.get()) {
            return null;
        }
        return queue.poll(timeoutMs, TimeUnit.MILLISECONDS);
    }

    /** Returns a message immediately if one is available, or {@code null} otherwise. */
    public MonitorMsg tryGetMonitorMessage() {
        if (closed.get()) {
            return null;
        }
        return queue.poll();
    }

    /** Returns {@code true} if {@link #stop()} or {@link #close()} has been called. */
    public boolean isClosed() {
        return closed.get();
    }

    /** Stops monitoring and releases resources. Idempotent. */
    public void stop() {
        if (closed.compareAndSet(false, true)) {
            GlideNativeBridge.closeMonitorClient(nativeMonitorId);
        }
    }

    @Override
    public void close() {
        stop();
    }

    // ---- helpers ----

    private static byte[] serializeConfig(GlideClientConfiguration config) {
        ConnectionRequest.Builder builder = ConnectionRequest.newBuilder();

        for (glide.api.models.configuration.NodeAddress addr : config.getAddresses()) {
            builder.addAddresses(
                    NodeAddress.newBuilder().setHost(addr.getHost()).setPort(addr.getPort()).build());
        }

        boolean insecureTls = TlsConfigHelper.resolveInsecureTls(config);
        if (config.isUseTLS()) {
            builder.setTlsMode(insecureTls ? TlsMode.InsecureTls : TlsMode.SecureTls);
        } else {
            builder.setTlsMode(TlsMode.NoTls);
        }

        byte[] rootCerts = TlsConfigHelper.extractRootCertificates(config);
        if (rootCerts != null) {
            builder.addRootCerts(ByteString.copyFrom(rootCerts));
        }

        ServerCredentials creds = config.getCredentials();
        if (creds != null) {
            AuthenticationInfo.Builder authBuilder = AuthenticationInfo.newBuilder();
            if (creds.getUsername() != null) {
                authBuilder.setUsername(creds.getUsername());
            }
            if (creds.getPassword() != null) {
                authBuilder.setPassword(creds.getPassword());
            }
            builder.setAuthenticationInfo(authBuilder.build());
        }

        if (config.getDatabaseId() != null) {
            builder.setDatabaseId(config.getDatabaseId());
        }

        if (config.getProtocol() != null) {
            if ("RESP2".equals(config.getProtocol().name())) {
                builder.setProtocol(ProtocolVersion.RESP2);
            } else if ("RESP3".equals(config.getProtocol().name())) {
                builder.setProtocol(ProtocolVersion.RESP3);
            }
        }

        return builder.build().toByteArray();
    }

    /**
     * Parse a JSON array string like {@code ["a","b","c"]} into a {@link List}.
     *
     * <p>Uses a simple hand-rolled parser to avoid a JSON library dependency.
     */
    static List<String> parseArgsJson(String json) {
        if (json == null || json.equals("[]")) {
            return Collections.emptyList();
        }
        List<String> result = new ArrayList<>();
        // Strip surrounding brackets
        int i = 1;
        int len = json.length();
        while (i < len - 1) {
            // Skip whitespace / commas
            char c = json.charAt(i);
            if (c == ',' || c == ' ') {
                i++;
                continue;
            }
            if (c == '"') {
                // Read quoted string
                i++; // skip opening quote
                StringBuilder sb = new StringBuilder();
                while (i < len) {
                    char ch = json.charAt(i++);
                    if (ch == '"') {
                        break;
                    }
                    if (ch == '\\' && i < len) {
                        char escaped = json.charAt(i++);
                        switch (escaped) {
                            case '"':
                                sb.append('"');
                                break;
                            case '\\':
                                sb.append('\\');
                                break;
                            case 'n':
                                sb.append('\n');
                                break;
                            case 'r':
                                sb.append('\r');
                                break;
                            case 't':
                                sb.append('\t');
                                break;
                            case 'u':
                                // Parse 4 hex digits (JSON unicode escape)
                                if (i + 4 <= len) {
                                    String hex = json.substring(i, i + 4);
                                    sb.append((char) Integer.parseInt(hex, 16));
                                    i += 4;
                                }
                                break;
                            default:
                                sb.append(escaped);
                        }
                    } else {
                        sb.append(ch);
                    }
                }
                result.add(sb.toString());
            } else {
                i++;
            }
        }
        return Collections.unmodifiableList(result);
    }
}

File: python/glide-sync/glide_sync/monitor_client.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

import json
import threading
import time
from typing import Callable, List, Optional

from glide_shared._glide_ffi import GlideFFI
from glide_shared.commands.core_options import MonitorMsg
from glide_shared.config import GlideClientConfiguration


class MonitorClient:
    """
    A client that streams all commands processed by the server via the MONITOR command.

    Must be used with a standalone (non-cluster) configuration.

    Warning: MONITOR is a debugging tool with performance implications.
    Do not use in production environments.
    """

    def __init__(self) -> None:
        self._ffi = GlideFFI.ffi
        self._lib = GlideFFI.lib
        self._core_client = self._ffi.NULL
        self._callback_ref = None
        self._queue: List[MonitorMsg] = []
        self._lock = threading.Lock()
        self._condition = threading.Condition(self._lock)
        self._is_closed = False
        self._user_callback: Optional[Callable[[MonitorMsg], None]] = None

    @classmethod
    def create(
        cls,
        config: GlideClientConfiguration,
        callback: Optional[Callable[[MonitorMsg], None]] = None,
    ) -> "MonitorClient":
        """
        Create a new MonitorClient connected to the server.

        Args:
            config: Standalone client configuration (must be GlideClientConfiguration).
            callback: Optional callback invoked for each MonitorMsg. If None, messages
                      are queued and retrievable via get_monitor_message().

        Returns:
            A MonitorClient instance.
        """
        if not isinstance(config, GlideClientConfiguration):
            raise TypeError(
                "MonitorClient requires a GlideClientConfiguration (standalone only)"
            )
        instance = cls()
        instance._user_callback = callback
        conn_req = config._create_a_protobuf_conn_request(cluster_mode=False)
        conn_req_bytes = conn_req.SerializeToString()

        @instance._ffi.callback("MonitorCallback")
        def _monitor_callback(
            client_ptr,
            timestamp,
            db,
            client_addr_ptr,
            client_addr_len,
            command_ptr,
            command_len,
            args_json_ptr,
            args_json_len,
        ):
            try:
                client_addr = bytes(
                    instance._ffi.buffer(client_addr_ptr, client_addr_len)
                ).decode("utf-8", errors="replace")
                command = bytes(instance._ffi.buffer(command_ptr, command_len)).decode(
                    "utf-8", errors="replace"
                )
                args_json_str = bytes(
                    instance._ffi.buffer(args_json_ptr, args_json_len)
                ).decode("utf-8", errors="replace")
                try:
                    args: List[str] = (
                        json.loads(args_json_str) if args_json_len > 0 else []
                    )
                except (json.JSONDecodeError, ValueError):
                    args = []
                msg = MonitorMsg(
                    timestamp=timestamp,
                    db=db,
                    client_addr=client_addr,
                    command=command,
                    args=args,
                )
                if instance._user_callback is not None:
                    instance._user_callback(msg)
                else:
                    with instance._condition:
                        instance._queue.append(msg)
                        instance._condition.notify()
            except Exception:
                pass  # Suppress callback errors to avoid crashing the Rust FFI layer

        instance._callback_ref = _monitor_callback
        client_response = instance._lib.create_monitor_client(
            conn_req_bytes, len(conn_req_bytes), _monitor_callback
        )
        if client_response == instance._ffi.NULL:
            raise RuntimeError("Failed to create monitor client: null response")
        if client_response.connection_error_message != instance._ffi.NULL:
            error = instance._ffi.string(
                client_response.connection_error_message
            ).decode()
            instance._lib.free_connection_response(client_response)
            raise RuntimeError(f"Failed to create monitor client: {error}")
        instance._core_client = client_response.conn_ptr
        instance._lib.free_connection_response(client_response)
        return instance

    def get_monitor_message(
        self, timeout: Optional[float] = None
    ) -> Optional[MonitorMsg]:
        """
        Block until a MonitorMsg is available, then return it.

        Args:
            timeout: Optional timeout in seconds. Returns None on timeout or if closed.
        """
        deadline = time.monotonic() + timeout if timeout is not None else None
        with self._condition:
            while not self._queue and not self._is_closed:
                remaining = None
                if deadline is not None:
                    remaining = deadline - time.monotonic()
                    if remaining <= 0:
                        break
                self._condition.wait(timeout=remaining)
            return self._queue.pop(0) if self._queue else None

    def try_get_monitor_message(self) -> Optional[MonitorMsg]:
        """Non-blocking retrieval. Returns None if no message is available."""
        with self._condition:
            return self._queue.pop(0) if self._queue else None

    def close(self) -> None:
        """Stop monitoring and release resources."""
        with self._condition:
            if self._is_closed:
                return
            self._is_closed = True
            self._condition.notify_all()
        if self._core_client != self._ffi.NULL:
            client = self._core_client
            self._core_client = self._ffi.NULL
            self._lib.close_monitor_client(client)
        self._callback_ref = None

    def stop(self) -> None:
        """Alias for close(). Stop monitoring and release resources."""
        self.close()

    def __enter__(self) -> "MonitorClient":
        return self

    def __exit__(self, *args) -> None:
        self.close()
